libs/structwalk/walktype.go

```diff
@@ -48,11 +48,21 @@ func walkTypeValue(path *structpath.PathNode, typ reflect.Type, visit VisitTypeF
 		return
 	}
 
-	kind := typ.Kind()
+	// Call visit on all nodes except the root node. We call visit before
+	// dereferencing pointers to ensure that the visit callback receives
+	// the actual type of the field.
+	if !path.IsRoot() {
+		visit(path, typ)
+	}
+
+	// Dereference pointers.
+	for typ.Kind() == reflect.Pointer {
```

Comment:
I explicitly dereference pointers here to avoid `visit` being called on both `*field` and `field` during recursive walk.